questionpy_sdk/webserver/app.py

```diff
@@ -57,18 +57,49 @@ async def render_options(request: web.Request) -> web.Response:
 
 @routes.post('/submit')
 async def submit_form(request: web.Request) -> web.Response:
-    """Get the options form definition and the form data on."""
+    """Store the form_data from the Options Form in the StateStorage."""
     webserver: 'WebServer' = request.app['sdk_webserver_app']
     form_data = await request.json()
+
+    async with webserver.worker_pool.get_worker(webserver.package, 0, None) as worker:
+        form_definition, _ = await worker.get_options_form(None)
+        parsed_form_data = webserver.state_storage.parse_form_data(form_definition, form_data)
+        try:
+            question = await worker.create_question_from_options(old_state=None, form_data=parsed_form_data)
+        except WorkerUnknownError:
+            return HTTPBadRequest()
+
+    question_state = json.loads(question.question_state)
+    webserver.state_storage.insert(webserver.package, question_state)
+
+    return web.json_response(question.question_state)
+
+
+@routes.post('/repeat')
+async def repeat_element(request: web.Request) -> web.Response:
+    """Add Repetitions to the referenced RepetitionElement and store the form_data in the StateStorage."""
+    webserver: 'WebServer' = request.app['sdk_webserver_app']
+    data = await request.json()
+    form_data = data['form_data']
+    repetition_reference = data['element-name'].replace(']', '').split('[')
+
     async with webserver.worker_pool.get_worker(webserver.package, 0, None) as worker:
+        manifest = await worker.get_manifest()
         form_definition, _ = await worker.get_options_form(None)
-        options = webserver.state_storage.parse_form_data(form_definition, form_data)
+        parsed_form_data = webserver.state_storage.parse_form_data(form_definition, form_data)
+        parsed_form_data = add_repetition(parsed_form_data, form_definition, repetition_reference)
         try:
-            question = await worker.create_question_from_options(old_state=None, form_data=options)
+            question = await worker.create_question_from_options(old_state=None, form_data=parsed_form_data)
         except WorkerUnknownError:
             return HTTPBadRequest()
 
     question_state = json.loads(question.question_state)
     webserver.state_storage.insert(webserver.package, question_state)
 
-    return web.json_response(form_data)
+    context = {
+        'manifest': manifest,
+        'options': form_definition.dict(),
+        'form_data': webserver.state_storage.get(webserver.package)
```

Comment:
Hier (und oben) sollten die aus `get_options_form` zurückgegebenen Form-Daten (zweites Element des Tupels) verwendet werden. Die sind möglicherweise anders als der State.

Reply:
`get_options_form` nimmt allerdings ein `Path` Objekt und keinen `str`, so wie der question_state vorliegt. 
Hast du eine Idee wie man das löst?

Reply:
Vermutlich müsstest du dann
1. Den bisherigen State laden, falls vorhanden
2. `get_options_form` damit aufrufen, um an die `form_definition` zu kommen
3. Dein `parse_form_data` mit der `form_definition` aufrufen, um an die Form-Daten zu kommen
4. Mit dem bisherigen State und den Form-Daten `create_question_from_options` aufrufen, um den neuen State zu bekommen
5. Den neuen State abspeichern
6. Mit dem neuen State erneut `get_options_form` aufrufen, um an eine (potenziell andere!) `form_definition` und die anzuzeigenden Form-Daten zu kommen
7. Rendern

Den ersten `get_options_form`-Call (also 2.) könntest du dir sparen, wenn du es schaffst, dass `parse_form_data` ohne die `form_definition` auskommt. Das geht auch, da du das im Grunde auf `_unflatten` reduzieren kannst, wenn du es so anpasst, dass Zahlen-Indexe (`[abc][0]`) dafür sorgen, dass Listen gebaut werden (für Repetitions).

Reply:
Ok habe ich jetzt so gemacht. 
Vielleicht kann es da aber zu Fehlern kommen, wenn man seinen Formelementen rein numerische Namen gibt. Ohne die `form_definition` und ohne das `kind` (text_input, static_text, etc.) im request zu übergeben, geht es glaube ich aber nicht anders.

Reply:
Danke :+1: 
Auf der DSL-Ebene werden rein numerische Namen implizit nicht erlaubt, weil der Feldname den Element-Namen angibt. Macht wahrscheinlich Sinn, wenn wir in `_BaseElement` aber noch eine Beschränkung nachziehen